mojo/proposals/string-ascii-unicode-utf-grapheme.md

```diff
@@ -0,0 +1,261 @@
+# String, ASCII, Unicode, UTF, Graphemes
+
+- Edit 0: 2025-02-05. Created.
+- Edit 1: 2025-04-27. Reduced the scope to only affect `StringSlice` and fixed
+  some details after feedback.
+- Edit 1: 2025-05-11. Reduced the scope even further and separated the work into
+  several much smaller phases.
+
+The current proposal will attempt to unify the handling of the standards.
+Providing nice ergonomics, as Python-like defaults as possible, and keeping the
+door open for optimizations.
+
+## String
+
+String is currently an owning type which encodes string data in UTF-8, which is
+a variable length encoding format. Unicode Codepoints can be 1 byte up to 4
+bytes long. ASCII text, which is where English text falls, is 1 byte long. As
+such, the defaults should optimize for that character set given it is most
+typical on the internet and on backend systems.
+
+## ASCII
+
+Ascii is pretty optimizable, there are many known tricks. The big problem
+with supporting only ASCII is that other encodings can get corrupted.
+
+## UTF standards
+
+- UTF-8 is 1-4 sets of 8 bits long
+- UTF-16 is 1-2 sets of 16 bits long
+- UTF-32 is 1 set of 32 bits long
+
+### When slicing by unicode codepoint e.g. "🔥🔥" (\U0001f525\U0001f525)
+
+- UTF-8: 8 sets (8 bytes) long. The first byte of each fire can be used to
+  know the length, and the next bytes are what is known as a continuation byte.
+  There are several approaches to achieve the slicing, they can be explored with
+  benchmarking later on.
+- UTF-16: 4 sets long. It's very similar in procedure to UTF-8.
+- UTF-32: It is fastest since it's direct index access. This is not the case
+  when supporting graphemes.
+
+## Graphemes
+
+Graphemes are an extension which allows unicode codepoints to modify the end
+result through concatenation with other codepoints. Visually they are shown as a
+single unit e.g. é is actually comprised of `e` followed by `´`  (\x65\u0301).
+
+Graphemes are more expensive to slice because one can't use faster algorithms,
+only skipping each unicode codepoint one at a time and checking the last byte
+to see if it's an extended cluster *.
+
+*: extended clusters is the full feature complete grapheme standard. There
+exists the posibility of implementing only a subset, but that defeats the
+purpose of going through the hassle of supporting it in the first place.
+
+## Context
+
+C, C++, and Rust use UTF-8 with no default support for graphemes.
+
+### Swift
+
+Swift is an interesting case study. For compatibility with Objective-C they went
+for UTF-16. They decided to support grapheme clusters by default. They recently
+changed the preferred encoding from UTF-16 to UTF-8. They have a `Character`
+type which is a generic representation of a Character in any encoding, that can
+be from one codepoint up to any grapheme cluster.
+
+### Python
+
+Python currently uses UTF-32 for its string type. So the slicing and indexing
```

Comment:
I don't think this is true anymore. They use the format outlined in PEP393, which can be UTF-8, UTF-16, or UTF-32 or some combination (new representations are created as-needed as copies). Also I think it has some way of preserving bogus codepoints.

Python strings are immutable, which makes the PEP393 approach easier.